honeycombio/data_source_query.go

```diff
@@ -284,3 +284,71 @@ func validateQueryJSON(validators ...querySpecValidateDiagFunc) schema.SchemaVal
 		return diagnostics
 	}
 }
+
+func extractFilters(d *schema.ResourceData) ([]honeycombio.FilterSpec, error) {
+	filterSet := d.Get("filter").(*schema.Set)
+	hashFn := filterSet.F
+	filters := make([]honeycombio.FilterSpec, filterSet.Len())
+	for i, schemaFilter := range filterSet.List() {
+		honeyFilter, err := extractFilter(d, hashFn(schemaFilter))
+		if err != nil {
+			return nil, err
+		}
+		filters[i] = honeyFilter
+	}
+	return filters, nil
+}
+
+const multipleValuesError = "must choose one of 'value', 'value_string', 'value_integer', 'value_float', 'value_boolean'"
+
+func extractFilter(d *schema.ResourceData, id int) (honeycombio.FilterSpec, error) {
+	var filter honeycombio.FilterSpec
+	filter.Column = d.Get(fmt.Sprintf("filter.%d.column", id)).(string)
+	filter.Op = honeycombio.FilterOp(d.Get(fmt.Sprintf("filter.%d.op", id)).(string))
+
+	valueSet := false
+	v, vOk := d.GetOk(fmt.Sprintf("filter.%d.value", id))
+	if vOk {
+		filter.Value = v
+		valueSet = true
+	}
+	vs, vsOk := d.GetOk(fmt.Sprintf("filter.%d.value_string", id))
+	if vsOk {
+		if valueSet {
+			return filter, fmt.Errorf(multipleValuesError)
+		}
+		filter.Value = vs
+		valueSet = true
+	}
+	vi, viOk := d.GetOk(fmt.Sprintf("filter.%d.value_integer", id))
+	if viOk {
+		if valueSet {
+			return filter, fmt.Errorf(multipleValuesError)
+		}
+		filter.Value = vi
+		valueSet = true
+	}
+	vf, vfOk := d.GetOk(fmt.Sprintf("filter.%d.value_float", id))
+	if vfOk {
+		if valueSet {
+			return filter, fmt.Errorf(multipleValuesError)
+		}
+		filter.Value = vf
+		valueSet = true
+	}
+	vb, vbOk := d.GetOk(fmt.Sprintf("filter.%d.value_boolean", id))
+	if vbOk {
+		if valueSet {
+			return filter, fmt.Errorf(multipleValuesError)
+		}
+		filter.Value = vb
+		valueSet = true
+	}
+
+	if filter.Op == honeycombio.FilterOpExists || filter.Op == honeycombio.FilterOpDoesNotExist {
+		if filter.Value != nil {
```

Comment:
This actually became a little nicer, we can do an actual nil check instead of just seeing if it matches the empty value.

Reply:
Indeed, this is much better 🙂